0.13.1 backports

laanwj opened the pull request:
Remaining backports for 0.13.1.

Trivial cherry-picks, except minor conflict in #8393:

```
error: could not apply 6aa28ab... Use cmpctblock type 2 for segwit-enabled transfer
```

This conflict is due to CConnMan changes.

sipa committed:
Use cmpctblock type 2 for segwit-enabled transfer

Contains version negotiation logic by Matt Corallo and bugfixes by
Suhas Daftuar.

Github-Pull: #8393
Rebased-From: 6aa28abf53ef4694692474b4a3b0a8fa7559b50b

jl2012 committed:
Make test framework produce lowS signatures

Github-Pull: #8499
Rebased-From: 9f0397aff7afa2afa9328daea0a2053122e79d44

theuni committed:
net: fix a few cases where messages were sent rather than dropped upon disconnection

75ead758 turned these into crashes in the event of a handshake failure, most
notably when a peer does not offer the expected services.

There are likely other cases that these assertions will find as well.

Github-Pull: #8862
Rebased-From: 905bc68d05595f41cca36b3df83accd10c00cc48

sdaftuar committed:
[qa] Add support for compactblocks v2 to mininode

Github-Pull: #8393
Rebased-From: 422fac649f75c907cad6ab7e2768b9032b9eae42

sdaftuar committed:
[qa] Update p2p-compactblocks.py for compactblocks v2

Github-Pull: #8393
Rebased-From: 27acfc1d2ee53cc52b54befd2d4bfa24a77a2eef

sdaftuar committed:
[qa] Fix bug in mininode witness deserialization

Also improve tx printing

Github-Pull: #8393
Rebased-From: f5b9b8f437c040205896ad0d7a6656efa08b5601

jnewbery committed:
Update bitcoin-tx to output witness data.

Github-Pull: #8817
Rebased-From: 4408558843c6c2b7abeb4160f641dfdbf5be5eb4

jl2012 committed:
Add policy: null signature for failed CHECK(MULTI)SIG

Github-Pull: #8634
Rebased-From: e41bd449ab2b8d01260795383af2c40b659d8587

jl2012 committed:
Make non-minimal OP_IF/NOTIF argument non-standard for P2WSH

Github-Pull: #8526
Rebased-From: c72c5b1e3bd42e84465677e94aa83316ff3d9a14

jl2012 committed:
Add NULLDUMMY verify flag in bitcoinconsensus.h

Github-Pull: #8848
Rebased-From: 2fa0063c26c80c719a1c0d30e548e338689ac917

jl2012 committed:
Require compressed keys in segwit as policy and disable signing with uncompressed keys for segwit scripts

Github-Pull: #8499
Rebased-From: 4c0c25a604cec39675187507df423cff5d707ab5

TheBlueMatt committed:
Use cmpctblock type 2 for segwit-enabled transfer

Contains version negotiation logic by Matt Corallo and bugfixes by
Suhas Daftuar.

Github-Pull: #8393
Rebased-From: 6aa28abf53ef4694692474b4a3b0a8fa7559b50b

TheBlueMatt committed:
Fix overly-prescriptive p2p-segwit test for new fetch logic

Github-Pull: #8393
Rebased-From: be7555f0c03057bb5537cc42ca9d4937389f0670

sdaftuar committed:
[qa] Add tests for uncompressed pubkeys in segwit

Github-Pull: #8499
Rebased-From: b811124202152424109b8e95ebe7ac25ff2e83c0

dagurval committed:
[qa] Fix compact block shortids for a test case

Github-Pull: #8904
Rebased-From: 4cdece40419bcc97345357f9268e03b0aff400b5

sdaftuar commented on qa/rpc-tests/p2p-compactblocks.py:
```diff
@@ -314,57 +366,63 @@ def test_compactblock_construction(self):
-                absolute_indexes = self.test_node.last_getblocktxn.block_txn_request.to_absolute()
+                assert(test_node.last_getblocktxn is not None)
+                absolute_indexes = test_node.last_getblocktxn.block_txn_request.to_absolute()
             assert_equal(absolute_indexes, [0])  # should be a coinbase request
```
Looks to be a bug in the test where the block manufactured in the test is of a too-low version, due to soft-fork activations in the regtest chain.  Only seems to manifest itself in the 0.13 branch because we hardcoded the soft-fork activation points in master to be at block heights that are slightly higher than the test, I believe.

Quick fix would be to explicitly set the block versions generated in this test to be version 4, I think, haven't tried yet...

TheBlueMatt committed:
[qa] Send segwit-encoded blocktxn messages in p2p-compactblocks

Github-Pull: #8916
Rebased-From: 032e883b937a6b70d5c367fc2ee57d7eea8cb473

jl2012 committed:
test segwit uncompressed key fixes

Github-Pull: #8499
Rebased-From: 9260085377e89e666d0cb95b462261d7e3a9c82f

TheBlueMatt committed:
[qa] Build v4 blocks in p2p-compactblocktests

This fixes an issue in backporting to 0.13 as 0.13 enforces SF
activation by block version lockin instead of through a hard-coded
block height.

Github-Pull: #8916
Rebased-From: a4ad37d4ef4bcd81bc9d867b277efdebc86bc2e8

TheBlueMatt commented:
#8922 (built on the now-merged #8904) should fix the above test failures.

jl2012 committed:
Add standard limits for P2WSH with tests

Github-Pull: #8499
Rebased-From: 3ade2f64cfe43ab53e4869ffc35d5fd23201e1c1

sipa committed:
Fix ismine and addwitnessaddress: no uncompressed keys in segwit

Github-Pull: #8499
Rebased-From: 248f3a76a825a332e5495c5947ad283a9e5e938f

jl2012 committed:
remove redundant tests in p2p-segwit.py

Github-Pull: #8499
Rebased-From: 67d6ee1e3679504f46473fe0818970565ff3b137

laanwj commented:
Added #8922 and #8499 - all applied cleanly except 9260085377e89e666d0cb95b462261d7e3a9c82f in the tests which also imported faaec1377c6975a7c427bac958b67a0aba9e1694
